07-13-16-language-plugins.md

```diff
@@ -0,0 +1,94 @@
+
+### tarBundle
+```.js
+// This must implement a Promise that resolves with a Buffer
+// that represents a DIRECTORY containing the desired bundle.
+exportables.tarBundle = function(options) {
+  // Should return a Promise which eventually gets resolved with the bundle
+  return Promise.resolve(bundle);
+};
+```
+`tarBundle` is a required step which should resolve a `Promise` with a Buffer that represents the tarballed code or binary. The CLI will automatically untar the bundle when after it is deployed on the Tessel 2.
+
+In the case of Rust, this was the biggest addition. In order to compile Rust code for Tessel, we need to set up a special cross compiler that can compile a binary that runs on Tessel's MIPS architecture. The easiest solution for now is to have the CLI send out the project to a preconfigured remote [cross compilation server](https://github.com/tessel/rust-compilation-server). The server then compiles and returns the binary to the CLI.
+
+### preRun
+```.js
+// Optionally make changes on Tessel before starting the program
+exportables.preRun = function(tessel, options) {
+  // Should return a Promise which eventually gets resolved
+  return Promise.resolve();
+};
+```
+`preRun` is an optional step that's useful if you need to call any shell functions on Tessel itself before the project is initialized. For example, with Rust, we set the permissions of the binary to ensure it's executable. It's unused for JavaScript.
+
+### shell/binary
+```
+exportables.binary = 'YOUR_EXECUTION_METHOD';
+exportables.shell = (options) => {
+  return tags.stripIndent `
+    #!/bin/sh
+    exec YOUR_EXECUTION_METHOD /app/remote-script/${opts.resolvedEntryPoint}
+  `;
+};
+```
+`binary` is a string that represents the engine that runs the program. It is used as the first argument in a shell command to start the program when a project is *`run`*. For JavaScript, this property is `node` and for Rust it's `.`.
+
+`shell` is a function that could have been named `begin` because it's responsible for actually getting the program going when you *`push`* the project. The reason it's called `shell` instead is because the function needs to return a string representing the contents of a shell script that starts the program.
+Tessel 2 operates via a series of bash commands and the CLI deployment process will write `langauge.shell` to a shell script file that automatically gets executed when Tessel boots up (by `init.d` for those familiar with Unix).
+
+### postRun
+```.js
+// Optionally make changes on Tessel after starting the program
+exportables.postRun = function(tessel, options) {
+  // Should return a Promise which eventually gets resolved
+  return Promise.resolve();
+};
+```
+`postRun` is another optional step that's useful if there is code you need to run on Tessel after the program has started. It's largely unused thus far so far but was added for future-proofness.
```

Comment:
That's correct, just after it starts the program, this function is called. The program may or may not have ended (I'll add that detail).